### ISSUE TITLE
Block Bindings: Support binding to `innerBlocks`

### ISSUE THREAD

**Issue by talldan** [WordPress/gutenberg#61225]
Note: this issue was originally written with the idea of supporting `innerBlocks` bindings for pattern overrides, but the scope has been broadened to all blocks. The description has been simplified.

## What problem does this address?
Some opportunities that arise from supporting bindings to innerBlocks have already been mentioned in #41073.

For pattern overrides, this would also be a hugely beneficial feature. Users could create synced patterns that have free areas for adding content. For example, imagine patterns that have particular locked design elements mixed with free written content. Pattern creators could also control the blocks that are insertable into A user would be able to control the design across every usage of that pattern.

It'd also make it possible to replace the Post Content block with a group that has innerBlocks bindings. Patterns would start to become as powerful as templates, and perhaps there's some convergence of those concepts.

## What is your proposed solution?

Previously block bindings have only supported binding attributes, but if bindings could also be configured for inner blocks, the above problems could be solved.

**Comment by fabiankaegy** [comment]
I really like the idea of this. Currently, we often have to build custom block wrappers that then have an open inner blocks area. Pattern Overrides in general still have the limitation, due to it's usage of `contentOnly` locking, of not even allowing the addition of new list items / buttons as described in: https://github.com/WordPress/gutenberg/issues/52018

**Comment by talldan** [comment]
> Pattern Overrides in general still have the limitation, due to it's usage of contentOnly locking, of not even allowing the addition of new list items / buttons as described in: https://github.com/WordPress/gutenberg/issues/52018

Yep, I think this is one area that makes me hesitant at the same time. I think there could still be separate solutions for blocks like list/buttons, those containers that have one supported inner block type (gallery and social links are maybe others) and aren't really freeform.

I think this might be good so that `contentOnly` can still be leveraged (but with additional features like adding/removing list items). That would allow creators to do things like setting a list style that can't be modified, or maybe setting the min/max number of blocks.

I would be hesitant of too many solutions for this kind of thing, but at the same time I think a freeform area in synced patterns would be quite a powerful feature.

**Comment by carlomanf** [comment]
> * Interestingly, I think it means patterns start encroaching on templates, the post content block works similarly to what's described above.

Over at #37753 a good point was made about "template tag" blocks being facsimiles of generic blocks (e.g. Post Title and Site Title being facsimiles of Heading) that would be better implemented as block bindings within the generic blocks. I guess the point quoted above is the same point regarding Post Content as a facsimile of Group.

Accordingly, the solution described here could enable post content data to be a block binding source for Group, and would be a better solution to #41073 which I would be happy to close in favour of this issue.

**Comment by youknowriad** [comment]
Great idea indeed. For me this means "inner blocks" support for block bindings. It's important that the implementation of such feature is not pattern overrides specific and would allow any block bindings source to inject blocks.

**Comment by eric-michel** [comment]
This is the holy grail of hybrid template / freeform editing and is something I've been dreaming of since the block editor was introduced. I agree with @youknowriad that making `InnerBlocks` support bindings is the way to go, which opens the door for many other block types to activate the same functionality (including custom blocks).

**Comment by talldan** [comment]
Another good use case for this will be nesting patterns. For example you could have a section that contains a header, images and an overridden grid/columns, and within the grid/columns you could nest card patterns. That allows for pretty cool composition of patterns.

**Comment by mtias** [comment]
With the ability to now express `allowedBlocks` directly in the markup this makes a lot of sense. I can see serializing these as nested elements. Curious if naming the group blocks is the only thing needed to now where to allocate these if you have more than one group block.

Side note, but we need to expose `allowedBlocks` in the Advanced UI of the group block as well, it's a super powerful feature. The UI can be similar to the block manager.

**Comment by talldan** [comment]
I've also been thinking on the idea of composing patterns. An example use case I have in my head is a pricing table pattern. It has a 'freeform' columns block, and by default there are a few columns, and each has a `pricing-table-card` pattern within it showing some placeholder content (name, features, pricing).

A challenge would be that when the user appends an additional column, the new column is going to be empty whereas I think it could be populated with the `pricing-table-card` pattern. The user could fuss around with finding that pattern  in the inserter, or duplicate the column, but I think it'd be better for the appender to insert the right content or suggest the right content for the column by default.

This use case isn't limited to only pricing tables, it's illustrative, there are plenty of situations where a pattern designer would want users to insert consistent content into columns, grids and other types of lists.

Some kind of `allowedPatterns` or `suggestedPatterns` API might work, or perhaps even allow pattern slugs in `allowedBlocks`. It could also be an entire category of patterns that's allowed/suggested.

This might be a few steps ahead of where we are right now, but it's good to think about how these things can be configured.

**Comment by Faton1995** [comment]
Hello, I would say my suggestion also fits here. 

https://github.com/WordPress/gutenberg/issues/63878

**Comment by gziolo** [comment]
Quoting the issue created by @liviopv:
- https://github.com/WordPress/gutenberg/issues/59606

> ## What problem does this address?
> For the Buttons Block (and eventually other blocks like List, Group, Column, Cover, Gallery), users might want to create patterns where all the nested blocks can be overridden, not just specific nested items. There should be a setting inside the parent block give users this option
> 
> ## What is your proposed solution?
> * Add a "Allow overrides in all nested blocks" option for blocks that only accept one type of nested block like Buttons, List, and Gallery
> * Add a "Allow overrides in nested blocks that support it" in blocks that can accept multiple types of blocks, like Group, Column and Cover.

It might be a good idea to consolidate these two issues and expand the scope for this one.

**Comment by jptweb** [comment]
Yes, I agree—this is the holy grail. I often face the challenge of creating multiple synced patterns for slight variations like “Hero Area With Text and a List” versus “Hero Area With Text, a List, and a Button.” An innerBlock (or a classic/WYSIWYG block with override support) would solve this.

Prioritizing this feature would let us immediately benefit from synced patterns even as broader block support is gradually added. It’s a win-win: developers can avoid building custom React blocks every time, and non-coders (who aren't going to be creating native blocks) can create consistent across site structured content without advanced HTML/CSS skills.

I did find some plugin developers are looking into this too and equally frustrated it being so close to possible [https://wordpress.org/support/topic/blocks-with-custom-block-editor-content/](https://wordpress.org/support/topic/blocks-with-custom-block-editor-content/)

I’m relatively new to modern WordPress (after using it in 2005 with a long hiatus) and will be testing interim solutions; wish I could help more with this, but also hear that its hardcoded in what blocks area allowed to have overrides; but if I can think of something to help in the meantime I will post. I hope its ok I mention this here, not sure if this the right place to simply +1 an idea but I see so much potential in this.

**Comment by JiveDig** [comment]
>Yes, I agree—this is the holy grail.

💯 This would be amazing and unlock so many amazing possibilities that would benefit agencies, theme/plugin shops, and ultimately the end user.

This has been a pain point that makes us bail and find alternative solutions (like a custom block, custom code, etc.).

I sort of see this as the block equivalent of shortcodes with closures. 

Inner blocks in (partially?) synced patterns FTW.

**Comment by vyskoczilova** [comment]
Hi, I've jumped into this issue since I couldn't find out how to get a card block where a user can add several paragraphs of content. Then I found out that it's not possible. Do you know how we can help move this forward? I could imagine allowing certain types of repetition of the override block (1-3 paragraphs) or setting the override to a group block, which would mean a wildcard for any content inside to be edited.

**Comment by tonnysantana2004** [comment]
This feature would be perfect! I was needing to develop this component:

<img width="250" height="240" alt="Image" src="https://github.com/user-attachments/assets/c30e6fd0-e1ff-40b4-8f93-897d91e03da4" />

---
I've been thinking on creating a pattern, and allowing the user to override the list group adding new items. So, I made this: 

```
add_filter(
    'block_bindings_supported_attributes_gowp/list-group',
    function ( $supported_attributes ) {
        $supported_attributes[] = 'content';
        $supported_attributes[] = 'innerBlocks';
        return $supported_attributes;
    }
);
```
But as innerBlocks are not supported for block bindings, this does not work. 

---

### So I've created a block for making this possible:

<img width="1911" height="945" alt="Image" src="https://github.com/user-attachments/assets/ff585bf2-1422-4f47-ba4b-0cd6dee8e225" />

---

However, it is not that cool having 20 blocks in my sidebar for very complex components, like cards, or sections, only because they need dynamic inner blocks.

It would be way better keeping the complex components as patterns, and simple ones such as buttons, lists etc, as blocks.

**Comment by cbravobernal** [comment]
I added `list-item` as a compatible block. It contains a fix in the HTML render to process the inner HTML present on a list, that way they work with nested items.
 - https://github.com/WordPress/gutenberg/pull/79346

Feel free to test. Meanwhile I'm trying to define the spects - issues for the innerBlocks.

**Comment by cbravobernal** [comment]
Related supporting PR: #79855 extracts the editor-side block-bindings context assembly into an internal getBlockBindingsContext helper.

This is useful for the innerBlocks bindings work because the innerBlocks resolver needs to call binding sources with the same source-visible context shape as attribute bindings: the block type usesContext plus the binding source usesContext, limited to context keys actually provided by ancestors. Keeping that logic shared avoids drift between attribute bindings and the future innerBlocks binding path.
